crates/egui/src/viewport.rs

```diff
@@ -0,0 +1,348 @@
+use std::{fmt::Display, sync::Arc};
+
+use epaint::Pos2;
+
+use crate::{Context, Id};
+
+/// This is used to send a command to a specific viewport
+///
+/// This is returned by `Context::get_viewport_id` and `Context::get_parent_viewport_id`
+#[derive(Default, Debug, Hash, Clone, Copy, PartialEq, Eq)]
+pub struct ViewportId(pub(crate) u64);
+
+impl Display for ViewportId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl ViewportId {
+    /// This will return the `ViewportId` of the main viewport
+    pub const MAIN: Self = Self(0);
+}
+
+/// This will deref to `ViewportIdPair::this`
+#[derive(Default, Debug, Hash, Clone, Copy, PartialEq, Eq)]
+pub struct ViewportIdPair {
+    pub this: ViewportId,
+    pub parent: ViewportId,
+}
+
+impl ViewportIdPair {
+    /// This will return the `ViewportIdPair` of the main viewport
+    pub const MAIN: Self = Self {
+        this: ViewportId::MAIN,
+        parent: ViewportId::MAIN,
+    };
+
+    pub fn new(this: ViewportId, parent: ViewportId) -> Self {
+        Self { this, parent }
+    }
+}
+
+impl std::ops::Deref for ViewportIdPair {
+    type Target = ViewportId;
+
+    fn deref(&self) -> &Self::Target {
+        &self.this
+    }
+}
+
+/// This is used to render an async viewport
+pub type ViewportRender = dyn Fn(&Context) + Sync + Send;
+
+pub type ViewportRenderSyncCallback = dyn for<'a> Fn(&Context, ViewportBuilder, ViewportIdPair, Box<dyn FnOnce(&Context) + 'a>)
+    + Send
+    + Sync;
+
+/// The filds in this struct should not be change directly, but is not problem tho!
+/// Every thing is wrapped in Option<> indicates that thing should not be changed!
+#[derive(PartialEq, Eq, Clone)]
+#[allow(clippy::option_option)]
+pub struct ViewportBuilder {
+    pub id: Id,
+    pub title: String,
+    pub name: Option<(String, String)>,
+    pub position: Option<Option<Pos2>>,
+    pub inner_size: Option<Option<Pos2>>,
```

Comment:
What is the difference between `None` and `Some(None)` ? Can we avoid using an Option of an Option?

Reply:
First when you call `Context::create_viewport` the builder will patch the existing builder of that viewport if exists, and will make `ViewportCommand` for the changes to be applied if a native window already exists!

For example like in android the app suspends and resumes or the compositor crashes on linux
on suspend the windows is destroyed or is already destroyed
on resume will build a new window and the window will be build with the ViewportBuilder that was patched
this make possible for a window to be restored and the and the ui is not be affected!

The difference between None and Some(None)
- None indicate that not change is needed
- Some(None) on inner_size will set the default native window size to be chosen by the OS when is rebuild!

This system can be removed if you want!